src/dev.rs

```diff
@@ -835,90 +835,128 @@ impl<F: FieldExt> MockProver<F> {
-                                })
-                                .unwrap_or_else(|| LookupFailure::OutsideRegion {
-                                    row: input_row as usize,
-                                });
+                            // Include zero rows in `inputs` only if `table` has no zero rows.
+                            if !table_has_zeros || t != zeros {
+                                // Also keep track of the original input row, since we're going to sort.
+                                Some((t, input_row))
+                            } else {
+                                None
+                            }
+                        })
+                        .collect();
+                    inputs.sort_unstable();
 
-                            Some(VerifyFailure::Lookup {
-                                lookup_index,
-                                location,
+                    let calc_input_columns = || {
+                        lookup
+                            .input_expressions
+                            .iter()
+                            .flat_map(|expression| {
+                                expression.evaluate(
+                                    &|_| vec![],
+                                    &|_| {
+                                        panic!("virtual selectors are removed during optimization")
+                                    },
+                                    &|index, _, _| vec![self.cs.fixed_queries[index].0.into()],
+                                    &|index, _, _| vec![self.cs.advice_queries[index].0.into()],
+                                    &|index, _, _| vec![self.cs.instance_queries[index].0.into()],
+                                    &|a| a,
+                                    &|mut a, mut b| {
+                                        a.append(&mut b);
+                                        a
+                                    },
+                                    &|mut a, mut b| {
+                                        a.append(&mut b);
+                                        a
+                                    },
+                                    &|a, _| a,
+                                )
                             })
-                        }
-                    })
+                            .collect::<HashSet<_>>()
+                    };
+
+                    let mut i = 0;
+                    inputs
+                        .iter()
+                        .filter_map(move |(input, input_row)| {
+                            let input_row = *input_row;
+                            while i < table.len() && &table[i] < input {
+                                i += 1;
+                            }
+                            if i == table.len() || &table[i] > input {
+                                assert!(table.binary_search(input).is_err());
+
+                                let input_columns = calc_input_columns();
```

Comment:
This doesn't depend on the input row so we could cache it, but that's a separate optimization, and would only affect the case where we have lots of lookup failures.